### PR TITLE
An Overview of Compiler Design

### LINKED ISSUE #1272: An Overview of Compiler Design
### Brief Summary: 
A compiler design is a computer program that transforms a source language (high-level language) into a machine language (low level language) without changing the program meaning. This article provides an overview of this concept. 

### Key Takeaways: 
1. Introduction to language processing system using compiler
2. The architecture of compiler design
3. Phases of a compiler
4. Applications of compiler technology
5. Conclusion

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by mbaabuones**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1272 .

### CODE REVIEW

**Review by cyberShaw — CHANGES_REQUESTED** [review]
Great article, @mbaabuones! I've suggested some minor changes. Please implement them and we can move on to final review.

File: articles/an-overview-of-compiler-design/index.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+Computers are important elements in our lives because they run applications that solve many real-life problems. Computer programmers play the role of writing codes (using programming languages) for computer programs to perform their tasks. However, the high-level programming language used can only be understood by human beings, but not the computer machine. A compiler is required to convert this high-level language to a language that can be comprehended by the machine.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:1]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
+Computers are important elements in our lives because they run applications that solve many real-life problems. Computer programmers play the role of writing codes (using programming languages) for computer programs to perform their tasks. However, the high-level programming language used can only be understood by human beings, but not the computer machine. A compiler is required to convert this high-level language to a language that can be comprehended by the machine. 
+
+In this article, we learn about the fundamental aspects of compiler design. Some of the aspects covered include language processing systems, compiler design architecture, and phases of a compiler. The article will also discuss the real-life applications of compiler technology.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:3]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
+Computers are important elements in our lives because they run applications that solve many real-life problems. Computer programmers play the role of writing codes (using programming languages) for computer programs to perform their tasks. However, the high-level programming language used can only be understood by human beings, but not the computer machine. A compiler is required to convert this high-level language to a language that can be comprehended by the machine. 
+
+In this article, we learn about the fundamental aspects of compiler design. Some of the aspects covered include language processing systems, compiler design architecture, and phases of a compiler. The article will also discuss the real-life applications of compiler technology.
+
+### What is a Compiler?
+A compiler is a computer program that transforms a source language (high-level language (HLL)) into a machine language (low-level language (LLL)) without changing the program meaning. It is simply an intermediary between the machine-readable language and the human-readable language. The principles of compiler design give an overview of the translation and optimization processes.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:6]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+A compiler can perform various operations such as parsing, preprocessing, lexical analysis, and semantic analysis. It can also perform code generation and code optimization. These operations are implemented in phases that consist of inputs and outputs.
+
+There are various types of compilers. Some of the common types include:
+* **Source-to-source compiler:** This is a compiler in which the source code of one language is transformed into a source code of a different language.
+* **Cross compiler:** In this compiler, the source code can be produced in one machine and executed in a different machine. 
+* **JIT (just in time) compiler:** In this compiler, the compilation is deferred until the runtime. It is applied to modern programming languages such as Python, Java, and JavaScript. 
+* **Hardware compiler:** This is a compiler that uses the hardware configuration to produce output, rather than a string of instructions.
+
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:15]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+![Language Processing System](/engineering-education/an-overview-of-compiler-design/language-processing-system.jpg)
+
+[Image Source: Tutorials Point](https://www.tutorialspoint.com/compiler_design/images/language_processing_system.jpg)
+
+The computer programmers first write the source code using an HLL. The source code then goes through the various components of the language processing system. These components are organized in a logical order. 
+
+The following provides a brief explanation of these components:
+* **Preprocessor:** This tool produces an output that is used as the input for the compiler. It performs various operations such as macro-processing, language extension, and file inclusion. 
+* **Compiler:** This compiles the high-level language and translates it into a language that can be understood by the assembler (assembly code or low-level language). 
+* **Assembler:** This tool uses the output of the compiler as its input. In this tool, the assembly code is transformed into a machine code. The output produced by the assembler can be referred to as an object file.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:28]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+[Image Source: Tutorials Point](https://www.tutorialspoint.com/compiler_design/images/language_processing_system.jpg)
+
+The computer programmers first write the source code using an HLL. The source code then goes through the various components of the language processing system. These components are organized in a logical order. 
+
+The following provides a brief explanation of these components:
+* **Preprocessor:** This tool produces an output that is used as the input for the compiler. It performs various operations such as macro-processing, language extension, and file inclusion. 
+* **Compiler:** This compiles the high-level language and translates it into a language that can be understood by the assembler (assembly code or low-level language). 
+* **Assembler:** This tool uses the output of the compiler as its input. In this tool, the assembly code is transformed into a machine code. The output produced by the assembler can be referred to as an object file. 
+* **Linker:** This tool transforms the output of the assembler into an executable machine code. Here, all the program parts are linked to enhance execution.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:29]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+* **Cross compiler:** In this compiler, the source code can be produced in one machine and executed in a different machine. 
+* **JIT (just in time) compiler:** In this compiler, the compilation is deferred until the runtime. It is applied to modern programming languages such as Python, Java, and JavaScript. 
+* **Hardware compiler:** This is a compiler that uses the hardware configuration to produce output, rather than a string of instructions.
+  
+### Introduction to Language Processing System 
+Learning the language processing system can improve our understanding of how a compiler is utilized. This system consists of various components that process the input language to produce the desired output. 
+
+![Language Processing System](/engineering-education/an-overview-of-compiler-design/language-processing-system.jpg)
+
+[Image Source: Tutorials Point](https://www.tutorialspoint.com/compiler_design/images/language_processing_system.jpg)
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:21]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+The compiler design architecture can be divided into two main parts: analysis and synthesis. 
+
+![Compiler Design Architecture](/engineering-education/an-overview-of-compiler-design/compiler-design-architecture.jpg)
+
+[Image Source: Scanftree](https://scanftree.com/compiler-design/images/compiler_analysis_synthesis.jpg)
+
+#### Analysis 
+This part represents the front-end in compiler design. It consists of various operations such as analyzing the source code, dividing the core into sections, and checking for errors.  It also constructs a symbol table to map source code symbols to relating information such as type, scope, and location. 
+
+An internal representation (IR) of the program is generated and analyzed before it is sent to the synthesis phase. The analysis part of the architecture consists of phases such as preprocessing, lexical analysis, syntax analysis, and semantic analysis.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:42]: Yes it is. Thank you for the correction.

File: articles/an-overview-of-compiler-design/index.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+[Image Source: The Tech Pro](http://www.kttpro.com/wp-content/uploads/2017/02/Compiler-1024x792.png)
+
+The following is a brief description of these phases.
+
+* **Lexical analysis:** This is the first phase of the compiler that receives the source code, scans, and transforms it into *lexemes*. These *lexemes* are represented by the lexical analyzer in a *token* form. *Tokens* consist of various categories such as separators, identifiers, operators, comments, and keywords. 
+* **Syntax analysis:** This phase is also referred to as *parsing*. It uses the tokens generated in the previous phase to produce a syntax tree (parse tree). It checks whether the token expressions are syntactically correct. 
+* **Semantic analysis:** This phase checks whether the language rules in the parse tree have been followed. Semantic information is added to the parse trees produced in the syntax analysis. The semantic analysis phase performs various operations such as checking for errors, associating variables with corresponding definitions, and issuing warnings. The output of this phase is in the form of annotated tree syntax.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:58]: Done.

File: articles/an-overview-of-compiler-design/index.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+The following is a brief description of these phases.
+
+* **Lexical analysis:** This is the first phase of the compiler that receives the source code, scans, and transforms it into *lexemes*. These *lexemes* are represented by the lexical analyzer in a *token* form. *Tokens* consist of various categories such as separators, identifiers, operators, comments, and keywords. 
+* **Syntax analysis:** This phase is also referred to as *parsing*. It uses the tokens generated in the previous phase to produce a syntax tree (parse tree). It checks whether the token expressions are syntactically correct. 
+* **Semantic analysis:** This phase checks whether the language rules in the parse tree have been followed. Semantic information is added to the parse trees produced in the syntax analysis. The semantic analysis phase performs various operations such as checking for errors, associating variables with corresponding definitions, and issuing warnings. The output of this phase is in the form of annotated tree syntax. 
+* **Intermediate code generation:** This phase involves generating an intermediate code that can be translated into the final machine code.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/an-overview-of-compiler-design/index.md:59]: Done.

**Comment by mbaabuones** [comment]
@cyberShaw I have finished working on the changes.

**Comment by cyberShaw** [comment]
@mbaabuones Please add a hero image to this.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by mbaabuones** [comment]
> 
> 
> @mbaabuones Please add a hero image to this.

@cyberShaw Done.

**Review by cyberShaw — APPROVED** [review]

**Comment by cyberShaw** [comment]
@mbaabuones Hero image should be in JPG. I've converted it and uploaded properly in jpg in commit: 466feec. Please ensure that you do it properly next time.

**Comment by mbaabuones** [comment]
> 
> 
> @mbaabuones Hero image should be in JPG. I've converted it and uploaded properly in jpg in commit: [466feec](https://github.com/section-io/engineering-education/commit/466feec5489c46fbb8682dd239a1a761fc9e4427). Please ensure that you do it properly next time.

@cyberShaw Noted. Thank you for the correction.

**Comment by hectorkambow** [comment]
> > @mbaabuones Hero image should be in JPG. I've converted it and uploaded properly in jpg in commit: [466feec](https://github.com/section-io/engineering-education/commit/466feec5489c46fbb8682dd239a1a761fc9e4427). Please ensure that you do it properly next time.
> 
> @cyberShaw Noted. Thank you for the correction.

jpg is best but PNG also works as well. 

@mbaabuones  cc @cyberShaw  👍

**Comment by cyberShaw** [comment]
> > > @mbaabuones Hero image should be in JPG. I've converted it and uploaded properly in jpg in commit: [466feec](https://github.com/section-io/engineering-education/commit/466feec5489c46fbb8682dd239a1a761fc9e4427). Please ensure that you do it properly next time.
> > 
> > 
> > @cyberShaw Noted. Thank you for the correction.
> 
> jpg is best but PNG also works as well.
> 
> @mbaabuones cc @cyberShaw 👍

Understood 👍🏼

**Pull request merged**